### PR TITLE
RoleVerifier and LivePresence v2

### PR DESCRIPTION
**Pull request by huntj88**
# Added new method getClientInfo to ILiveShareHost. 
```typescript
getClientInfo(clientId: string): Promise<IClientInfo | undefined>;
```

This method provides a way for clients to query for `IClientInfo`, which contains information about themselves or other meeting participants.

```typescript
export interface IClientInfo {
    userId: string;
    roles: UserMeetingRole[];
    displayName?: string;
}
```

In previous versions of live-share, properties like `userId` and `displayName` were provided by the user and sent over live-share-events. This new api is necessary to prevent spoofing of `userId` and `displayName`, by having the clients independently retrieve that information about other clients.

### LivePresence and LiveCanvas have been updated to use this new method.
* LivePresence now prevents presence highjacking by reporting a different userId and displayName than your actual one.
* LiveCanvas no longer allows a user to provide a displayName, and automatically uses the one provided by the `ILiveShareHost`

### Replace getClientRoles with getClientInfo
On older teams clients, calling `getClientRoles` is expensive. In the `BackwardsCompatibilityHostDecorator` we retain a high cache time of 5 minutes for `getClientRoles`. `getClientInfo` is cheap to call and is designed to be called frequently.

### Other changes
* Moved network request caching from role verifier to host decorator
* Added a backwards compatibility host decorator
* Added static method to `LiveShareClient`
  - getClientInfo
* Moved static methods from `LiveEvent` to `LiveShareClient`
  - getTimestamp
  - verifyRolesAllowed
  - setTimestampProvider
  - setRoleVerifier
* Updated samples
* Migrated from Lerna to NPM Workspaces
  - Simpler and faster
  - No static reference issues
  - Doesn't mask package.json issues if a package/sample isn't included in a package.json
* Changed some fluid-framework & live-share packages to peer dependency model
* Upgraded Fluid Framework versions to ^1.3.6 to fix various issues some developers ran into with Tinylicious
* Renamed/reorganized params for methods & deprecated old ones
  - `updatePresence(state, data)` ->`update(data, state)`. This aligns better with `LiveState` changes and also prioritizes data setting, which is more common than state setting.
  - `sendEvent(evt)` to `send(evt)`. This aligns better with `LiveState` and `LivePresence`.

### CODE REVIEW

File: packages/live-share-canvas/src/core/LiveCanvas.ts [hunk 1]
```diff
@@ -509,11 +518,11 @@ export class LiveCanvas extends DataObject {
             scope,
             InkingEventNames.pointerMove,
             (evt: IPointerMovedEvent, local: boolean) => {
-                if (!local && evt.clientId) {
+                const clientId = evt.clientId;
```
Comment (dclaux) [inline packages/live-share-canvas/src/core/LiveCanvas.ts:521]: Why this change? That's an unnecessary constant... I could understand if it was like `someLongVarName.eventLonger.somethingElse.clientId`, but `evt.clientId` is just as clear as `clientId`

File: packages/live-share-canvas/src/core/LiveCanvas.ts [hunk 2]
```diff
@@ -775,20 +784,26 @@ export class LiveCanvas extends DataObject {
...
-                const screenPosition =
-                    this._inkingManager.viewportToScreen(position);
-
-                liveCursor.setPosition(screenPosition);
-            } else {
-                this.removeCursor(clientId);
-            }
+        if (position) {
+            LiveShareClient.getClientInfo(clientId).then((clientInfo) => {
+                if (this._inkingManager) {
```
Comment (dclaux) [inline packages/live-share-canvas/src/core/LiveCanvas.ts:792]: Why not keep this as the first condition? It's cheap (I assume cheaper than `LiveShareClient.getClientInfo`) and necessary to run the code anyway.

File: packages/live-share-canvas/src/core/LiveCanvas.ts [hunk 3]
```diff
@@ -55,18 +56,29 @@ enum InkingEventNames {
...
     /**
      * Optional. The user's display name.
      */
     displayName?: string;
+}
+
+/**
+ * Encapsulates information about a user that is okay to be sent through events.
+ */
+export interface IEventUserInfo {
```
Comment (dclaux) [inline packages/live-share-canvas/src/core/LiveCanvas.ts:74]: I am not sure I understand the split of `IUserInfo` into two interfaces. Both `displayName` and `pictureUrl` are optional, IUserInfo is only used to instantiate a `LiveCursor` class, and `LiveCanvas` owns building the IUserInfo object, so it will always pass it the correct `displayName`. Unless I'm misreading your changes, the split isn't necessary and you can (should) simply keep the existing `IUserInfo` interface.

File: packages/live-share-canvas/src/core/LiveCanvas.ts [hunk 4]
```diff
@@ -425,8 +434,12 @@ export class LiveCanvas extends DataObject {
         liveStroke.clear();
     };
 
-    private getLocalUserInfo(): IUserInfo | undefined {
-        return this.onGetLocalUserInfo ? this.onGetLocalUserInfo() : undefined;
+    private getLocalEventUserInfo(): IEventUserInfo {
```
Comment (dclaux) [inline packages/live-share-canvas/src/core/LiveCanvas.ts:437]: Why change the name of this method? It's private anyway.

Comment (dclaux) [reply packages/live-share-canvas/src/core/LiveCanvas.ts:437]: This is where you should be calling `LiveShareClient.getClientInfo()` as the role of `getLocalUserInfo` is to build the `IUserInfo` object. You shouldn't do it in `updateCursorPosition`. If that means `getLocalUserInfo` needs to become async, that's fine.

**Review by dclaux — CHANGES_REQUESTED** [review]

File: packages/live-share-canvas/src/core/LiveCanvas.ts (hunk 4 above)
Comment (ryanbliss) [reply packages/live-share-canvas/src/core/LiveCanvas.ts:437]: In James's changes, we are not broadcasting the `displayName` in the socket event anymore, since that is untrustworthy & easily spoofed. `LiveShareClient.getClientInfo()` checks all clientId (local or non-local), so `getLocalUserInfo` isn't really the same thing. He separated them because `LiveShareClient.getClientInfo()` does not yet support profile pictures, and thus that still gets broadcasted through the socket (though there is an argument for removing that until it does, since it is easily spoofable).

File: packages/live-share-canvas/src/core/LiveCanvas.ts (hunk 2 above)
Comment (ryanbliss) [reply packages/live-share-canvas/src/core/LiveCanvas.ts:792]: isn't there a thing in typescript where if using something that is possibly undefined in an async function, that you have to do the if check again or it will error out, since it could technically be reset to undefined before the async op completes?

File: packages/live-share-canvas/src/core/LiveCanvas.ts (hunk 3 above)
Comment (ryanbliss) [reply packages/live-share-canvas/src/core/LiveCanvas.ts:74]: In James's changes, we are not broadcasting the `displayName` in the socket event anymore, since that is untrustworthy & easily spoofed. `LiveShareClient.getClientInfo()` checks all clientId (local or non-local), so `getLocalUserInfo` isn't really the same thing. He separated them because `LiveShareClient.getClientInfo()` does not yet support profile pictures, and thus that still gets broadcasted through the socket (though there is an argument for removing that until it does, since it is easily spoofable).

File: packages/live-share-canvas/src/core/LiveCanvas.ts (hunk 2 above)
Comment (dclaux) [reply packages/live-share-canvas/src/core/LiveCanvas.ts:792]: You might be right

File: packages/live-share-canvas/src/core/LiveCanvas.ts (hunk 4 above)
Comment (huntj88) [reply packages/live-share-canvas/src/core/LiveCanvas.ts:437]: also, the naming change was to make it clear how it was being used and what is being returned, which is now only a subset of `IUserInfo`

File: packages/live-share-canvas/src/core/LiveCanvas.ts (hunk 3 above)
Comment (dclaux) [reply packages/live-share-canvas/src/core/LiveCanvas.ts:74]: `onGetLocalUserInfo` is supposed to allow the developer to provide their own display name and picture url. With these changes, you will always ignore the `displayName` provided by the developer and use your own instead. Thus, that interface split and change in the definition of `IUserInfo` only serves as a way to seemingly not break the consuming app while you are actually breaking it if the developer is using `onGetLocalUserInfo` and is returning a custom `displayName`... in short, you're hiding a breaking change.

The right thing to do here is to introduce an explicit breaking change and force the developer to update their code.  There is no need to change `IUserInfo` because it already accdurately represents the data that needs to be transmitted via the event. You just need to change `onGetLocalUserInfo` so it only allows the developer to return a picture Url. Maybe simply change the event into `onGetLocalUserPicture` and have it return a string.

Or, we may even simply drop the ability to display a picture: why is it important to enforce a non-spoofable name while at the same time not enforce a non-spoofable picture?

**Review by dclaux — CHANGES_REQUESTED** [review]

File: packages/live-share-canvas/src/core/LiveCanvas.ts (hunk 4 above)
Comment (dclaux) [reply packages/live-share-canvas/src/core/LiveCanvas.ts:437]: Please see my comment below. I think you are just hiding a breaking change when you should explicitly introduce a breaking change that forces developers to update their code.

File: packages/live-share-canvas/src/core/LiveCanvas.ts (hunk 2 above)
Comment (dclaux) [reply packages/live-share-canvas/src/core/LiveCanvas.ts:789]: I'm only saying that ideally we don't do something expensive unless it's going to be useful, and here if `this._inkingManager` is undefined then `LiveShareClient.getClientInfo` will be called for nothing. Technically it's really not that big a deal because chances are `LiveCanvas` will always have an associated `InkingManager`, but still.

File: packages/live-share/src/internals/PolyfillHostDecorator.ts [hunk 5]
```diff
@@ -0,0 +1,220 @@
...
+    UserMeetingRole,
+    IClientInfo,
+} from "../interfaces";
+import { RequestCache } from "./RequestCache";
+import { waitForResult } from "./utils";
+
+const EXPONENTIAL_BACKOFF_SCHEDULE = [100, 200, 200, 400, 600];
+const CACHE_LIFETIME = 5 * 60 * 1000; // original cache time
+
+export interface InternalDontUseGetClientInfoRetryPolyfill {
```
Comment (ryanbliss) [inline packages/live-share/src/internals/PolyfillHostDecorator.ts:20]: Add @hidden to comment for docs

File: packages/live-share/src/internals/PolyfillHostDecorator.ts [hunk 6]
```diff
@@ -0,0 +1,220 @@
...
+     * For internal use only.
+     */
+    getClientInfo(
+        clientId: string,
+        retrySchedule?: number[]
+    ): Promise<IClientInfo | undefined>;
+}
+
+/**
+ * Experiment for polyfil
```
Comment (ryanbliss) [inline packages/live-share/src/internals/PolyfillHostDecorator.ts:33]: Add @hidden to comment for docs

File: packages/live-share-canvas/src/core/LiveCanvas.ts (hunk 2 above)
Comment (huntj88) [reply packages/live-share-canvas/src/core/LiveCanvas.ts:780]: discussed in a meeting, okay with changes, and getClientInfo is cheap to call

File: packages/live-share-canvas/src/core/LiveCanvas.ts (hunk 3 above)
Comment (huntj88) [reply packages/live-share-canvas/src/core/LiveCanvas.ts]: discussed in meeting, came up with a solution where IUserInfo extends IEventUserInfo, instead of IUserInfo being a type union.

File: packages/live-share-canvas/src/core/LiveCanvas.ts (hunk 4 above)
Comment (huntj88) [reply packages/live-share-canvas/src/core/LiveCanvas.ts]: resolved in meeting and other conversation thread

**Review by dclaux — APPROVED** [review]

File: packages/live-share-canvas/src/core/LiveCanvas.ts [hunk 7]
```diff
@@ -775,20 +772,26 @@ export class LiveCanvas extends DataObject {
...
-
-                liveCursor.setPosition(screenPosition);
-            } else {
-                this.removeCursor(clientId);
-            }
+        if (position) {
+            LiveShareClient.getClientInfo(clientId).then((clientInfo) => {
+                if (this._inkingManager) {
+                    const userInfo: IUserInfo = {
+                        displayName: clientInfo?.displayName,
```
Comment (ryanbliss) [inline packages/live-share-canvas/src/core/LiveCanvas.ts:782]: I've learned from one key partner that they have intentionally not implemented `displayName` when implementing the previous callback, opting just to use profile picture. Spoofing is less of a concern for them though they are considering that. In any case, might be worth adding a flag/setting to hide display names.

Comment (ryanbliss) [reply packages/live-share-canvas/src/core/LiveCanvas.ts:782]: Let's take this up in a separate PR
